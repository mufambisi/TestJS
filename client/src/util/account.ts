import {
  DigitalAccountStatus,
  DigitalAccountType,
} from "../graphql-types/rx-gateway/globalTypes";
import { UserAccounts_viewer_digital } from "../graphql-types/rx-gateway/UserAccounts";

export function hasActiveAccount(
  data: UserAccounts_viewer_digital,
  accountType: DigitalAccountType
) {
  return data.accounts.some(
    (account) =>
      account.type === accountType &&
      account.status === DigitalAccountStatus.ACTIVE
  );
}

export const getOpenSapAccounts = (data: UserAccounts_viewer_digital) => {
  return data.accounts.filter(
    (it) =>
      it.type === DigitalAccountType.SAP &&
      it.status === DigitalAccountStatus.ACTIVE
  );
};

export function hasAccountOnly(
  data: UserAccounts_viewer_digital,
  accountType: DigitalAccountType
) {
  return data.accounts.every(
    (account) =>
      account.type === accountType &&
      account.status === DigitalAccountStatus.ACTIVE
  );
}
