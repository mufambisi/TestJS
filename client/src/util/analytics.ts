import {
  errorForm,
  reportException,
  ICardAA,
  IInteractionCardData,
  interactionCard,
} from "@origin-digital/reporting-client";

export const trackError = (error: Error, formName?: string) => {
  reportException(error.message, error);
  errorForm({
    errorCode: error.name,
    errorMessage: error.message,
    formName,
  });
};

export const trackingProviderCaptureClickCard = (
  card: Pick<ICardAA, "id" | "title" | "primaryCta">
) => ({ label }: Pick<IInteractionCardData, "label">) => {
  const interactionCardData: IInteractionCardData = {
    card: {
      category: "dashboard-card",
      ...card,
    },
    label,
  };

  interactionCard("click", interactionCardData);
};
