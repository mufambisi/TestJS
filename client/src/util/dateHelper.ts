import { parseISO, isValid, format } from "date-fns";

export function parseDate(
  dateString: string | null | undefined
): Date | undefined {
  if (!dateString || dateString.trim().length === 0) {
    return undefined;
  }
  const parsedDate = parseISO(dateString);
  if (isValid(parsedDate)) {
    return parsedDate;
  } else {
    return undefined;
  }
}

export function formatDate(
  date: Date | undefined,
  formatString: string
): string {
  if (!date) {
    return "";
  }
  return format(date, formatString);
}
