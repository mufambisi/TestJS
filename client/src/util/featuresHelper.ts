import {
  FeatureStatus,
  FeatureType,
} from "../graphql-types/rx-gateway/globalTypes";
import {
  UserAccounts_viewer_digital,
  UserAccounts_viewer_digital_features,
} from "../graphql-types/rx-gateway/UserAccounts";

export function getFeature(
  data: UserAccounts_viewer_digital,
  featureType: FeatureType
): UserAccounts_viewer_digital_features | undefined {
  return data.features.find(
    ({ type, status }) =>
      type === featureType && status === FeatureStatus.ACTIVE
  );
}

export function hasFeature(
  data: UserAccounts_viewer_digital,
  featureType: FeatureType
) {
  return !!getFeature(data, featureType);
}
