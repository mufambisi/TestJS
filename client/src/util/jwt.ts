import {
  getOriginCustomerId,
  decodeJwtPayload,
  ClaimKey,
} from "@origin-digital/auth-utils";
import { fetchAccessToken } from "@origin-digital/event-dispatcher";

export async function getAccessToken() {
  const jwt = await fetchAccessToken({});
  return jwt?.accessToken ?? undefined;
}

export function getCustomerEmailFromJwt(payload: string): string {
  return decodeJwtPayload(payload)[ClaimKey.EMAIL];
}

export function getCustomerIdFromJwt(payload: string): string {
  const custId = getOriginCustomerId(payload) ?? "";
  return custId;
}

export async function getCustomerId(): Promise<string | undefined> {
  const accessToken = await getAccessToken();
  if (!accessToken) return;
  return getCustomerIdFromJwt(accessToken);
}
