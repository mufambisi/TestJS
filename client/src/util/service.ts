import { UserAccounts_viewer_digital } from "../graphql-types/rx-gateway/UserAccounts";
import {
  DigitalServiceStatus,
  DigitalServiceType,
} from "../graphql-types/rx-gateway/globalTypes";

export function hasActiveService(
  data: UserAccounts_viewer_digital,
  serviceType: DigitalServiceType
) {
  return data.services.some(
    ({ status, type }) =>
      type === serviceType && status === DigitalServiceStatus.ACTIVE
  );
}
