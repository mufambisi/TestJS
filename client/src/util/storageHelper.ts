import { differenceInCalendarDays } from "date-fns";

// expiry in days
export const storageExpiry = 90;

export type DashboardStorageKeys =
  | "missing-features-banner"
  | "lpg-missing-features-banner";
export interface StorageItemStatus {
  status: "closed" | "shown";
  dismissedAt?: Date;
}

export type DashboardLocalStorage = {
  [key in DashboardStorageKeys]?: StorageItemStatus;
};

const getDashboardKey = (customerId: string) => `dashboard-rx-${customerId}`;

export const getDashboardLocalStorage = (
  customerId: string | undefined,
  storageType: DashboardStorageKeys
): StorageItemStatus | undefined => {
  if (!customerId) {
    return;
  }

  const data = localStorage.getItem(getDashboardKey(customerId));
  if (!data) {
    return;
  }

  const parsedData: DashboardLocalStorage = JSON.parse(data);
  const statusInfo = parsedData?.[storageType];
  if (statusInfo?.dismissedAt) {
    if (
      differenceInCalendarDays(new Date(), new Date(statusInfo?.dismissedAt)) >
      storageExpiry
    ) {
      delete parsedData[storageType];
      localStorage.setItem(
        getDashboardKey(customerId),
        JSON.stringify(parsedData)
      );
      return;
    }
  }

  const itemData = parsedData?.[storageType];
  if (itemData) {
    return {
      status: itemData.status,
      dismissedAt: itemData.dismissedAt
        ? new Date(itemData.dismissedAt)
        : undefined,
    };
  }
  return undefined;
};

export const setDashboardLocalStorage = (
  customerId: string | undefined,
  storageType: DashboardStorageKeys,
  status: StorageItemStatus["status"] = "closed"
): void => {
  if (!customerId) {
    return;
  }

  const itemStatus: StorageItemStatus = {
    status,
  };
  if (status === "closed") {
    itemStatus.dismissedAt = new Date();
  }

  const rawData = localStorage.getItem(getDashboardKey(customerId));
  const data = rawData ? JSON.parse(rawData) : {};
  data[storageType] = itemStatus;

  localStorage.setItem(getDashboardKey(customerId), JSON.stringify(data));
};

export const hasItemBeenDismissed = (
  customerId: string | undefined,
  storageType: DashboardStorageKeys
) => getDashboardLocalStorage(customerId, storageType)?.status === "closed";

export const storeDismissedItem = (
  customerId: string | undefined,
  storageType: DashboardStorageKeys
) => setDashboardLocalStorage(customerId, storageType, "closed");
