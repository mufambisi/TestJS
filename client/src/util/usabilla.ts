import {
  getAccessToken,
  getCustomerIdFromJwt,
  getCustomerEmailFromJwt,
} from "./jwt";

async function fetchOriginCustomerDetailsFromJwt(): Promise<{
  originCustomerId?: string;
  originCustomerEmail?: string;
}> {
  const accessToken = await getAccessToken();
  if (!accessToken) return {};

  return {
    originCustomerId: getCustomerIdFromJwt(accessToken),
    originCustomerEmail: getCustomerEmailFromJwt(accessToken),
  };
}

export async function usabillaWebHandler() {
  window?.["usabilla_live"]?.("data", {
    custom: await fetchOriginCustomerDetailsFromJwt(),
  });
  window?.["usabilla_live"]?.("click");
}
