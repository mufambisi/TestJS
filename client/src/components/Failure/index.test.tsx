import React from "react";
import { render, fireEvent } from "@origin-digital/ods-testing-library";
import { trackError } from "src/util/analytics";
import { Failure } from "./index";

const id = "id";
const error = new Error("error");
const onClick = jest.fn();
const Component = () => <div data-id="component" />;

jest.mock("src/util/analytics");

const renderFailure = (condition?: Error) =>
  render(
    <Failure id={id} error={condition} onClick={onClick}>
      <Component />
    </Failure>
  );

describe("<Failure />", () => {
  it("should render without error", () => {
    const { container, getAllByTestId } = renderFailure();

    expect(getAllByTestId("component")).toHaveLength(1);
    expect(container).toMatchSnapshot();
  });

  it("should call onClick() when CTA is clicked", () => {
    const { getByTestId } = renderFailure(error);

    fireEvent.click(getByTestId("failure-cta"));

    expect(onClick).toHaveBeenCalled();
  });

  it("should call on trackError", () => {
    const { getByTestId } = renderFailure(error);

    fireEvent.click(getByTestId("failure-cta"));

    expect(trackError).toHaveBeenCalledWith(error, id);
  });
});
