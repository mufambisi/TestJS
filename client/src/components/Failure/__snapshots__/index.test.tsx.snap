// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Failure /> should render without error 1`] = `
<div>
  <div
    data-id="component"
  />
</div>
`;
